### ISSUE TITLE
[RFC] Imageinfo changes in 3.0

### ISSUE THREAD

**Issue by SahAssar** [bolt/bolt#4679]
The output of the imageinfo function has changed quite a bit in master, A bunch of "nice to have" data seems to have been removed, like aspectratio and landscape/portrait/square.

It would be nice if we could agree on what we want to include in it, and hopefully make it as compatible with 2.2 themes as possible.

Example of 2.2 output:

	array:12 [▼
	  "width" => 1391
	  "height" => 1852
	  "type" => "jpeg"
	  "mime" => "image/jpeg"
	  "aspectratio" => 0,75107991360691
	  "filename" => "2016-01/19142717931-30117fa7ef-o.jpg"
	  "fullpath" => "/var/www/test/files/2016-01/19142717931-30117fa7ef-o.jpg"
	  "url" => "/files/2016-01/19142717931-30117fa7ef-o.jpg"
	  "exif" => array:5 [▼
		"latitude" => ""
		"longitude" => false
		"datetime" => false
		"orientation" => false
		"aspectratio" => 0,75107991360691
	  ]
	  "landscape" => false
	  "portrait" => true
	  "square" => false
	]

Example of master/3.0 output:

	Info {#2418 ▼
	  #dimensions: Dimensions {#2095 ▼
		#width: 1391
		#height: 1852
	  }
	  #type: Type {#1856 ▼
		-id: 2
		-name: "JPEG"
	  }
	  #bits: 8
	  #channels: 3
	  #mime: "image/jpeg"
	  #exif: Exif {#2096 ▼
		#data: array:4 [▼
		  "FileSize" => 492026
		  "MimeType" => "image/jpeg"
		  "height" => 1852
		  "width" => 1391
		]
		#rawData: array:8 [▼
		  "FileName" => "9k="
		  "FileDateTime" => 0
		  "FileSize" => 492026
		  "FileType" => 2
		  "MimeType" => "image/jpeg"
		  "SectionsFound" => ""
		  "COMPUTED" => array:4 [▼
			"html" => "width="1391" height="1852""
			"Height" => 1852
			"Width" => 1391
			"IsColor" => 1
		  ]
		  "IPTC" => []
		]
	  }
	}

**Comment by rarila** [comment]
Especially bad because (IMHO) you can't expect EXIF data to be there. I assume it is not build, just read from the file?!

What I miss: Size should really be in. Filename and directory path (namespace?) would be nice, also.

Then aspect ratio… easily can be calculated, but having some algorythm returning a human friendly ``3:4`` out of ``0,75107991360691`` would be nice.

Another nice one would be a mapping to [fontawesome file icons](http://fontawesome.io/icons/#file-type), although that would be better one level up so that it also works for non image files.

And a last more general one (I may be wrong here, enlighten me :-): It looks like we only save something like {"file":"2foo\/bar.jpg"} for images. That seems to be like a bottleneck to me, as needed information then has to be extracted from the files e.g. for contentype listing thumbnails on every page request. At the absolute minimum I expect width and height to be saved in storage for performance reasons.

**Comment by SahAssar** [comment]
That example is from a file with basically no exif data, by the way. Personally I'd like some of the data to be in the "old" style, since it seems more user friendly and would aid theme compatibility.

Both aspect ratio and filesize would be nice to have in a human-readable format in addition to the current one, like @rarila said.

**Comment by rarila** [comment]
@SahAssar Ok, just tried with a PNG file… seems like generic EXIF data is generated. I think I would prefer having those fields as simple "old" style data and EXIF data only there if there is real EXIF data in the image file included. At least it would probably be interesting to have a flag somewhere that says there's real exif data inside the file or not.

**Comment by C4Grey** [comment]
Dont know for sure how it is realized now in the master branch, but would not it be better if retrieval of an image info will be moved to a service accessible via DI in the Application container(like db, slugify, etc.)? It allow us to use a single format of image data to store in database in our extensions and also it would be easier to replace this service by a custom implementation.

**Comment by GawainLynch** [comment]
Yes, very much so @C4Grey 

We're still waiting on comment from @CarsonF as to his intended design goals, but generally that is what we've been working towards.

**Comment by CarsonF** [comment]
@SahAssar All of that data is still there, it has just been moved to getters instead of predetermined values. This saves processing as it may not be used. The `Image\Info` class has ArrayAccess as well for BC.

If there is really a use case for `dump`ing that data, then I could implement `__debugInfo` on `Image` that would convert it back to an array when dumping.

@GawainLynch I think there is a mistake though as `ImageHandler::imageInfo` should return an `Image` not the `Image\Info` as this would include the path data as well (info is already magically forwarded to Info class for BC).

The only thing that is expected to change is `fullpath`. Instead of it returning the absolute path it returns the filesystem abstractions version of it. e.g. `files://foo/cat.gif`.

@rarila The struggle with caching that data is invalidation. @GawainLynch and I have gone back and forth on this several times. Reality is it is a hard problem and we agreed it is better to error on the side of accuracy rather than performance.

**Comment by CarsonF** [comment]
@C4Grey With the new filesystem abstraction this data can be pulled from anywhere. 
```php
$image = $app['filesystem']->get('files://foo/cat.gif');

// Info methods
$image->getInfo()->isLandscape(); // true
$image->getInfo()->getWidth(); // 600
// Info calls are forwarded from Image for BC
$image->getWidth(); // 600

// Path methods
$image->getFullPath(); // files://foo/cat.gif
$image->getFilename(); // cat.gif
$image->getFilename('.gif'); // cat

// Filesystem manipulation methods
$image->rename('super_awesome_cat.gif');
$image->delete();
```

I agree that image info handling would be better if it was DI'ed, but the goal here was for simplicity. Due to how Bolt is initialized the `filesystem` has to be loaded very early and trying to injected all those things would just be more for the user to do. But definitely open to the idea in a later pass.

**Comment by GawainLynch** [comment]
> I think there is a mistake though as `ImageHandler::imageInfo` should return an Image not the `Image\Info` as this would include the path data as well (info is already magically forwarded to Info class for BC).

Brain barely functioning right now… But I seem to remember us trying to return an `Image` and then falling back to `Image\Info` due to something BC related… can't… remember… what… though… :frowning:

**Comment by rarila** [comment]
The deeper problem that that makes one think about caching such things at all is one I have criticized quite a while ago. The strict separation of uploaded images (files) into filesystem and database entry that are only losely connected with a pathname. The only advantage of this is that you can use an image twice (which I believe is a seldom usecase).

That little advantage costs a big price. (To be on topic) no saving of image data with the entry as you always can change the image external. Every action inside the file manager (moving, renaming and deleting files) will most likely break stuff as there's absolutely no feedback where one file is used or if it is used at all. After deleting an entry you have to remember used files and delete them by hand taking quite some extra time, is error prone and impossible to do for the non techy user. And even for those it would be an impossible task as soon as the number of files or entries reaches some threshold. Managing files gets impossible as soon as there are multiple editors. You, having a big brain, probably can remember what files you have uploaded and where you have used them. But you can't know if your file is linked by another editors entry. So in the end file upload is a one-way road. FINO - first in, never out …

Sorry for my rant, I need this from time to time.

**Comment by CarsonF** [comment]
@rarila I totally understand. How do other people/products solve this problem? From what I've seen every CMS has this problem. Do you have any solutions?

**Comment by GDmac** [comment]
In another CMS i use there is a specific fieldtype called channelimages (and channelfiles). The files are stored in a "entryID" subfolder and thusly linked to the entry. The fieldtype ads listeners for delete entry to be able to remove unused files together with the entry. The fieldtype also allows linking to files from other entries. 

**Inside content**
The big caveat with any filemanagement is when someone starts using these images/files in a CKedit or markdown field. There are some possibilities that are being used in that other cms. For instance, on save, any imagepath or filepath is replaced with a placeholder, e.g. ``/files/20/foo.jpg`` placeholder like ``{{filedir}}20/foo.jpg``. This opens a route to search/replace stuff inside textfields, but is somewhat weak link.

**alternative inside content**
Another approach that is being used (in that other cms) is have the images in the image type field, and use only placeholders directly inside the CKedit or markdown. ``{{image1 class="rightfloat"}}`` which then are search/replaced directly on the front-end in the template. This restricts filemanagement to the file/image field and editors can simply reference the images inside the content.

:+1: for an option for managed images/files based on entryID (delete with entry)

**Comment by rarila** [comment]
@CarsonF I'm glad I finally found one (two) who also sees the problem.

I solved that in the past (way before markdown and wyswyg) by using placeholders in a self parsed format or the image wasn't part of editable content (e.g. product photos). The image path and image properties like width/height were saved in the same db entry. Deleteing the image, set the path to ``null`` and delete the file, Uploading a new file deleted the old one and renaming was also done in the normal editing form. There was no other access to the files accept hardcore ftp of course:-)

Ok, now I would probably add a new system table for resources (files, images videos) where every entry has an id and holds some standard properties, like width/height, size, video/audio playtime, … that could be useful for displaying or processing. That id would be linked (foreign keys, 1:1, n:1) to contenttype entries using them (image, file, imagelist, filelist, …). That way you can find out how often, where or if at all an image is used. There could even be rights be connected with the file db entry so that only the (right-) owner could use/delete/…the file. The filemanager would access the files through this db layer, so that no file that is still used can be deleted. As there's no path inside the contenttype entry, moving an renaming the file is not a problem. (You can see the file id like an inode number). Another thing could be done: the file could be saved (flat) with the id or some md5 like string as name and the filename/path from the request would be simply translated to the real path. That way you even could move resources from local storage to CDn or whatever without changing or breaking any ct entry.

Now for thte caveat. As @GDmac already pointed out, we now have html/markdown content that brings some complication. But I think, nothing that can't be handled. I guess it would invove some parsing and replacing on save (like @GDmac pointed out). Then on save you also could warn on non existent files. Every contentfield that could link to files could have another system field that links to the image resource … I think you get my idea.

Many ifs, coulds, hows, probablies and things to work out but I think it could be done with our great team. The thing I'm sure, overall usabilty and ease of use would benefit massively.

**Comment by SahAssar** [comment]
@CarsonF Ah, I see. I'm pretty used to dumping things like imageinfo to find out the structure of them and how to use them, but that could just as well be covered in the docs/cheatsheet :)

**Comment by CarsonF** [comment]
@rarila Can you open another ticket for your concerns?

**Comment by sibputnik** [comment]
hi. ImageInfo return values changed, but in docs we have old description. Please, fix it.
